[2단계 - 장바구니] 제나(이정은) 미션 제출합니다.

JeLee-river opened the pull request:
안녕하세요 쵸파! 2단계 장바구니 미션의 리뷰를 요청드립니다! 마지막까지 잘부탁드립니다. ㅎㅎ

**현재, 마감 기한내에 요구사항을 모두 반영하지 못해, 일부 기능이나 테스트가 동작하지 않는 문제가 있습니다.** 1차 요청 후, 피드백과 함께 반영하여 다시 요청드리겠습니다. 정말 죄송합니다. 😭

## 📦 장바구니 미션

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

### 1단계

- 클라이언트 상태를 효과적으로 모델링하고 관리할 수 있다.
- Jest, React Testing Library(RTL)를 활용하여 주요 기능에 대한 테스트를 작성할 수 있다.

### 2단계

- 상태를 중복 없이 구성하고, 계산이 가능한 값은 파생 상태로 도출하여 관리할 수 있다.
- 함수(컴포넌트, 훅, util)가 단일한 책임만 갖도록 분리하고, 그 책임에 맞는 명확한 인터페이스를 정의할 수 있다.

## 🕵️ 셀프 리뷰(Self-Review)

### 제출 전 체크 리스트

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 테스트 케이스를 모두 작성했나요?
- [x] 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 배포 링크 기입: [배포 링크](https://react-shopping-cart-git-step2-jeleerivers-projects.vercel.app/)


<br>

###  미구현된 요구사항

- 제출 기한을 지키고자, 요구사항을 완전히 충족하지 못한 상태에서 1차 리뷰를 요청드리게 되었습니다. 2차 요청 전까지 최대한 빠르게 반영하여 재요청드리도록 하겠습니다. 아래는, 요구사항에 명시되어 있으나, 구현하지 못한 기능 또는 현재 인지하고 있는 문제입니다. 

1. 사용자의 장바구니 선택을 로컬 스토리지로 관리
2. 새로 추가된 기능에 대한 테스트
3. 테스트 오류 현상
4. 전체 선택된 상태에서 장바구니 아이템 삭제 시, 전체 선택이 해제되는 문제
5. 쿠폰 선택 로직과 관련하여, undefined 가능성으로 인해 발생하는 타입 문제

<br>

### 리뷰 요청 & 논의하고 싶은 내용

### 1) 상태 관리 설계 의도

전반적으로 최상위 컴포넌트, 즉, 페이지 컴포넌트 단에서 상태를 선언하고 하위 컴포넌트들에 props로 전달하는 방식으로 구현하려 했습니다.
또한, 하위 컴포넌트에서 상위 컴포넌트의 상태를 변경해야 하는 경우, setState 함수를 직접 주지 않고, 이를 한 번 더 랩핑해서 핸들러 함수를 전달하는 방식으로 구현했습니다.

<br>

### 2) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

이번 단계에서 가장 많이 고민한 점은 OOP 원칙 준수입니다. 특히, useCheckboxHandler의 로직들을 리팩토링하면서, 이에 대해 가장 많이 고민했습니다.

1단계에서 구현한 useCheckboxHandler는 장바구니 아이템 전용으로 작성되어 있어, 다른 id 아이템 배열엔 그대로 쓸 수 없는 문제가 있었습니다. 따라서 범용 인터페이스로 추상화하고자, 한 곳에 작성되어 있던 코드를 '개별 선택', '전체 선택', '최대 선택 개수가 제한된 상태의 로직'으로 분리하고 파일을 나누었습니다. 이 과정에서 각 파일의 코드가 역할과 책임에 맞게 분리되도록 구현하려 하면서 시간이 오래 걸렸습니다.

<br>

### 3) 이번 리뷰를 통해 논의하고 싶은 부분

### 1. useCheckboxHandler의 options 매개변수

useCheckboxHandler가 options를 받도록 구현한 점이 오히려 useCheckboxHandler가 너무 많은 역할을 하도록 하고 있는지, 새로운 커스텀 훅을 만드는 것이 나을지 고민이 되어, 이 부분을 질문드리고 싶습니다.

현재 체크 박스 로직을 담당하는 useCheckboxHandler는 체크 박스를 활성화할 아이템 외에 option을 받고 있습니다. 이 option은 전체 체크 박스의 사용 여부, 전체 선택 활성화의 기본값 설정 여부, 최대 선택 제한 개수로 이루어져 있습니다.

이와 같이 구현한 이유는, 컴포넌트에 따라 체크박스를 사용하는 방식에 약간의 차이가 있더라도, 같은 로직을 재사용할 수 있도록 하고 싶었기 때문입니다. 그러나 이 과정에서 최대 선택 가능 개수를 제한한다면, 전체 선택을 할 수 없는데 이 둘을 고려해서 로직을 짜야할지 고민이 들었습니다. 또한 최대 선택 개수를 제한하고 전체 선택을 활성화 하는 경우는 어떻게 대처할 지에 대해서도 생각해보게 되었습니다.

따라서 validate라는 파일을 생성하여 최대 선택 개수를 제한했을 때 전체 선택를 활성화하려 하면 에러를 던지도록 구현했습니다.

그러나 이러한 방식이 복잡하다는 생각도 들어, 재사용성과 범용성을 고려하다가 useCheckboxHandler가 너무 커진 것은 아닐까 하는 생각이 들었습니다. 유사한 로직이라도 장바구니 아이템에 사용되는 선택 로직과, 쿠폰에 사용되는 선택 로직을 서로 분리하는 것이 좋을지 고민도 되었습니다. 이에 대한 쵸파의 의견이 궁금합니다..!

<br>

### 2. 상태 관리 및 컴포넌트 간 의존

현재 코드에서 사용되는 상태들이 적절한 곳에서 생성되어 관리되고 있는지 궁금합니다..!
시간상, 일단 코드가 돌아가도록 하는 것을 우선순위로 두고 코드를 구현하면서, 리팩토링을 필수적으로 해야하는 상황이 되었습니다. 😂
혹시 상태 관리 위치나 컴포넌트 간 의존 관계면에서 리팩토링에 참고하거나 개선되면 좋을 점이 있을지 궁금합니다!

<br>

### 3. 컴포넌트의 역할과 책임 분리

하나의 컴포넌트에서 지나치게 많은 역할을 수행하고 있거나 책임 분리가 필요한 곳이 있을지 궁금합니다!

---

## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도뿐만 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. 상태관리

- 원본상태/파생상태를 적절히 구분하여 선언하였나요?
- React state 를 불필요하게 선언한 부분은 없나요?
- 상태 관리 로직의 책임이 적절히 응집/분리되었나요? (ex. reducer, hook, Context)

### 2. 역할과 책임

- 단일 책임 원칙(SRP)에 맞게 컴포넌트/함수가 분리되었나요?
- 컴포넌트/함수의 인터페이스가 구현에 의존하지 않고, 책임에 맞게 설계되었나요?

### 3. 테스트

- 쿠폰 기능에 대한 다양한 시나리오가 충분히 테스트되었나요?
- 함수의 책임을 분리하고, 순수성을 유지하는 등 테스트하기 좋은 코드를 작성하였나요?

---

감사합니다!!

woowa-euijinkk committed:
chore: 의존성 제거

JeLee-river committed:
refactor: Description 컴포넌트를 범용적으로 사용할 수 있도록 리팩토링

JeLee-river committed:
refactor: 사용자가 장바구니 아이템을 변경하는 UI를 조건부로 보여주도록 리팩토링

JeLee-river committed:
refactor: 주문 확인 페이지가 Description 컴포넌트를 사용하도록 리팩토링

JeLee-river committed:
refactor: CartItemList 컴포넌트를 children을 감싼 레이아웃 담당 형태로 리팩토링

- 불필요하게 props가 거쳐가고 있기 때문에 children을 감싸주는 형태로 리팩토링

JeLee-river committed:
feat: 주문 확인 페이지에 CartItemList 컴포넌트 추가

JeLee-river committed:
feat: Modal 컴포넌트 구현

JeLee-river committed:
feat: 쿠폰 모달을 주문 확인 페이지에서 열 수 있도록 기능 구현

JeLee-river committed:
refactor: modal 컴포넌트의 위치를 components 폴더 내부로 이동

JeLee-river committed:
feat: 서버의 쿠폰 데이터를 실사용 타입으로 변환하는 함수 생성

JeLee-river committed:
feat: 쿠폰 데이터를 받아오는 api 함수 생성

JeLee-river committed:
feat: 서버의 시간 데이터를 서비스에 맞게 가공하는 유틸함수들을 구현

JeLee-river committed:
feat: 쿠폰 유형별 데이터 타입을 생성

JeLee-river committed:
feat: 서버의 쿠폰 데이터를 사용할 형식에 맞게 가공하는 함수 구현

JeLee-river committed:
feat: 시간을 문자열로 파싱하는 함수 추가

JeLee-river committed:
feat: Coupon 컴포넌트 생성

JeLee-river committed:
refactor: Coupon 컴포넌트가 쿠폰 타입에 따라 UI를 다르게 보여주도록 수정

JeLee-river committed:
feat: 쿠폰의 할인 유형에 맞게 데이터 변환 로직을 적용하는 함수 생성

JeLee-river committed:
feat: 쿠폰 모달에 쿠폰 목록이 렌더링되도록 코드 추가

JeLee-river committed:
refactor: Coupon 컴포넌트 내부에서만 사용되는 유틸 함수를 Coupon 디렉터리로 이동

JeLee-river committed:
refactor: 조건부로 렌더링되는 쿠폰 정보를 컴포넌트로 분리

JeLee-river committed:
refactor: Coupon 타입을 Coupon 컴포넌트 디렉터리에서 관리하도록 분리

JeLee-river committed:
feat: 쿠폰의 discountType에 따라 세부 정보를 렌더링하는 CouponDetails 컴포넌트 생성

JeLee-river committed:
style: 쿠폰아이템에 스타일 적용

JeLee-river committed:
feat: 쿠폰 목록을 렌더링하는 CouponList 컴포넌트 구현

JeLee-river committed:
style: 모달 스타일 추가

JeLee-river committed:
refactor: 아이템 리스트와 각 아이템의 태그를 ul, li로 변경

JeLee-river committed:
style: 모달 내부 컨텐츠에 스타일 적용

JeLee-river committed:
style: Modal 컴포넌트 스타일 수정

JeLee-river committed:
refactor: useCheckboxHandler 함수의 범용성 개선

- cartItems에 맞춰진 로직을 범용적으로 개선함

- 전체 선택 박스 활성화, 자동 전체 선택, 최대 선택 갯수를 옵션으로 받도록 리팩토링

JeLee-river committed:
refactor: useCheckboxHandler에서 단일 선택, 전체 선택 로직을 분리

- 책임 분리를 위해 단일 선택, 전체 선택 로직을 커스텀 훅으로 분리함

JeLee-river committed:
feat: 선택가능 아이템이 제한된 경우에 사용되는 로직을 관리하는 useLimitSelect 훅 생성

JeLee-river committed:
feat: options의 유효성을 검증하는 함수 추가

JeLee-river committed:
refactor: 전체 선택 로직의 가독성 개선

JeLee-river committed:
feat: 쿠폰에 체크박스 기능 추가

JeLee-river committed:
chore: Modal 컴포넌트의 폴더명이 대문자로 시작하도록 수정

JeLee-river committed:
refactor: 쿠폰 모달을 새로운 컴포넌트로 분리

JeLee-river committed:
refactor: 상품명의 태그를 h3에서 p로 변경

JeLee-river committed:
refactor: 페이지의 본문 태그를 section에서 main으로 변경

JeLee-river committed:
feat: 추가 배송 정보 컴포넌트 구현

JeLee-river committed:
refactor: 선택된 아이템 정보를 반환하는 함수를 유틸 파일로 이동

JeLee-river committed:
style: 불필요한 마진을 제거하고 요소 간 배치 간격 조정

JeLee-river committed:
fix: FixedCoupon 타입에 필드가 누락된 문제 수정

JeLee-river committed:
feat: 쿠폰 모달을 주문 확인 페이지에 연결

JeLee-river committed:
feat: 쿠폰 로직을 관리하는 useCoupon 커스텀 훅 생성

JeLee-river committed:
refactor: 쿠폰 적용이 불가한 조건인 경우, 쿠폰이 비활성화되도록 코드 추가

JeLee-river committed:
feat: 주문 확인 페이지의 결제 내역을 보여주는 OrderReceipt 컴포넌트 생성

JeLee-river committed:
feat: OrderComplete 페이지 생성

JeLee-river committed:
fix: 체크박스를 해제했을때 체크박스 UI가 사라지는 문제 수정

JeLee-river committed:
refactor: 주문 금액이 10만원 이상일 때, 추가 배송비도 무료가 되도록 수정

JeLee-river committed:
style: 장바구니 상품의 개수에 맞게 장바구니 리스트의 높이가 조절되도록 수정

JeLee-river committed:
test: useCheckboxHandler 코드 수정에 따른 테스트 수정

JeLee-river committed:
fix: 전체 선택을 체크한 상태에서 아이템 삭제 시, 선택 토글이 반대로 되는 문제 수정

JeLee-river committed:
refactor: 금액을 계산하는 로직을 유틸함수로 분리

JeLee-river committed:
refactor: 범용적으로 사용되는 컴포넌트를 commons 디렉터리로 분류

bassyu commented on src/hooks/checkbox/useCheckboxHandler.ts:
> useCheckboxHandler가 options를 받도록 구현한 점이 오히려 useCheckboxHandler가 너무 많은 역할을 하도록 하고 있는지, 새로운 커스텀 훅을 만드는 것이 나을지 고민이 되어, 이 부분을 질문드리고 싶습니다.

지금도 최대한 범용적으로 잘 만들어 주셨네요~!

두 방법 다 괜찮다고 생각하는데요!
차이점이 많아서, 여러 옵션간의 사이드 이팩트를 막고 싶다면 훅을 분리할 것 같고,
옵션이 사소해서 공통점이 더 많다면 옵션을 사용할 것 같아요.

지금 return type이나 동작에 차이점이 좀 있어서 분리 해보면 좋겠습니다!

JeLee-river replied:
useCheckboxHandler에서 전체 선택과 관련된 로직을 분리하여 useAllCheckboxHandler를 새롭게 생성했습니다!

[refactor: 전체 선택 체크박스과 관련된 로직을 useCheckboxHandler에서 분리](https://github.com/woowacourse/react-shopping-cart/pull/407/commits/2857bdbcfbbea9d87eb0177bd645e9b7dae3353e)

bassyu commented on src/pages/CouponModal/CouponModal.tsx:
```diff
@@ -0,0 +1,110 @@
+    fetchCoupons();
+  }, [fetchCoupons]);
+
+  const { selectedIds, toggleSelect, isSelected, isMaxSelected } =
```
타입 에러가 발생하고 있습니다!

JeLee-river replied:
[이전 코멘트](https://github.com/woowacourse/react-shopping-cart/pull/407#discussion_r2143694573)의 커밋에서  useCheckboxHandler의 반환 타입이 안정적으로 바뀌면서 타입 에러가 해결되었습니다!

bassyu commented on src/hooks/checkbox/useCheckboxHandler.ts:
> 1단계에서 구현한 useCheckboxHandler는 장바구니 아이템 전용으로 작성되어 있어, 다른 id 아이템 배열엔 그대로 쓸 수 없는 문제가 있었습니다. 따라서 범용 인터페이스로 추상화하고자, 한 곳에 작성되어 있던 코드를 '개별 선택', '전체 선택', '최대 선택 개수가 제한된 상태의 로직'으로 분리하고 파일을 나누었습니다. 이 과정에서 각 파일의 코드가 역할과 책임에 맞게 분리되도록 구현하려 하면서 시간이 오래 걸렸습니다.

시간이 오래 걸리는 이슈가 있었군요..!
그래서 많이 치명적이지 않다면, 결정을 빨리 하는 것이 중요한 것 같아요 (저도 연습 중인 부분입니다..)
코드도 중요하지만, 마감기한을 지키는 것이 우선순위가 높으니까요~!

우선 중복을 허용하며 최대한 빠르게 구현 하고 보면, 완전히 중복된 부분이 잘 보이더라구요!
애매한 중복이 있을 때, `빠른 중복 허용` -> `중복 제거` 순서로 구현하는 방법도 추천합니다~!

bassyu commented on src/hooks/useCoupon.ts:
알고보니 훅이 아니였네요~!
검증에 실패했을 때 에러를 throw하는데요, 이 기능이 네이밍에 잘 표현되면 좋겠습니다!

JeLee-river replied:
확실히 훅 보다는 유틸 함수에 가까운 로직들을 모아놓은 파일이였던 것 같습니다. 😂

기존에 쿠폰과 관련된 역할을 담당했던 useCoupon을 제거하고, useCoupon의 로직들을 `domain/couponService` 와 `domain/couponHandler`로 리팩토링했습니다!! couponHandler는 구현에 가까운 기능들을 모아두었고 couponService는 이 기능들을 호출하는 로직을 추상화하여 사용할 수 있도록 했습니다.

[refactor: 쿠폰 할인 금액 계산과 적용 가능한 쿠폰을 판별하는 기능을 추상화](https://github.com/woowacourse/react-shopping-cart/pull/407/commits/25b0bd737c0b4781f49f03a6105dc3c3280472e8)

bassyu commented on src/pages/CouponModal/CouponModal.tsx:
```diff
@@ -0,0 +1,110 @@
+  filterNonExpiredCoupons,
+  decideCanRedeem,
+}: CouponModalProps) {
+  const [coupons, setCoupons] = useState<CouponType[]>([]);
+  const { fetchData } = useFetch<CouponDataType[]>("coupons");
+  const fetchCoupons = useCallback(
+    () =>
+      fetchData({
+        apiCall: getCoupons,
+        onSuccess: (data) => {
+          if (data) {
+            setCoupons(data.map((coupon) => adaptCoupon(coupon)));
+          }
+        },
+        onError: (error) => {
+          const errorMessage =
+            error instanceof Error ? error.message : DEFAULT_ERROR_MESSAGE;
+          alert(errorMessage);
+        },
+      }),
+    [fetchData]
+  );
+
+  useEffect(() => {
+    fetchCoupons();
+  }, [fetchCoupons]);
```
이 묶음을 추상화 하면 어떤 장점이 있을까요?

JeLee-river replied:
호출부에서 구체적인 패칭 과정을 알 필요 없이, 패칭에 성공한 경우, 실패한 경우 실행하고자 하는 로직을 전달하여 필요한 데이터를 얻을 수 있는 장점이 있습니다! 또한 패칭 로직이 변경되더라도, 추상화된 코드와 호출부는 수정하지 않아도 되어, 유지보수 관점에서 좋다고 생각합니다.

bassyu reviewed (changes requested):
안녕하세요 제나!
2단계도 잘 부탁드립니다~!

미구현된 요구사항이 있었군요 😿 😿 
우선 간단하게 코멘트 드릴테니, 빠르게 구현을 마친 후 다시 요청 주세요!
기능도 보면서 같이 피드백 드리는 것이 좋을 것 같습니다~!

> 하나의 컴포넌트에서 지나치게 많은 역할을 수행하고 있거나 책임 분리가 필요한 곳이 있을지 궁금합니다!

단위는 잘 나눠주신 것 같아요~! 👍  조금 커보이는 곳에 코멘트 남겼습니다!
또는 제나가 특히 걱정되는 페이지나 컴포넌트가 있을까요?

---

마지막까지 파이팅입니다~! 💪 💪

JeLee-river committed:
refactor: 쿠폰 할인 금액 계산과 적용 가능한 쿠폰을 판별하는 기능을 추상화

- 기존에 useCoupon에 작성되어 있던 쿠폰과 관련된 로직을 유틸성 함수로 분리

- Service계층을 생성하여 쿠폰들의 공통 로직을 추상화

- handler를 사용하여, 쿠폰의 종류에 상관없이 동일한 방식으로 기능이 동작할 수 있도록 리팩토링

JeLee-river committed:
feat: 현재 시각이 특정 시간 범위 내에 존재하는지 확인하는 유틸 생성

JeLee-river committed:
refactor: 쿠폰 할인 금액이 음수로 표현되도록 수정

JeLee-river committed:
feat: 최적의 쿠폰 할인 조합을 적용하는 기능 추가

JeLee-river committed:
fix: 쿠폰 배열이 빈 배열일 때 할인금액이 -Infinity가 되는 문제 수정

JeLee-river committed:
feat: 사용자의 체크 박스 선택을 로컬 스토리지에 저장하는 기능 구현

JeLee-river committed:
refactor: 로컬 스토리지 기능 구현에 따른 코드 리팩토링

JeLee-river committed:
refactor: 주문 확인 페이지의 주문 상품목록을 컴포넌트화

- CartItem이 안정적으로 필요한 props를 받도록 하기 위해 OrderItem 컴포넌트를 새롭게 생성함

JeLee-river committed:
chore: 불필요한 주석 제거

JeLee-river committed:
fix: 처음 페이지에 접속했을 때 전체 선택이 해제되는 문제 수정

JeLee-river committed:
refactor: 전체 선택 체크박스과 관련된 로직을 useCheckboxHandler에서 분리

JeLee-river committed:
refactor: 콘솔창의 경고와 관련된 코드 수정

JeLee-river committed:
test: 로컬 스토리지 도입에 따른 테스트 코드 수정

JeLee-river committed:
test: 주문 확인 페이지임을 체크하는 테스트 로직 수정

JeLee-river committed:
test: 쿠폰 테스트 추가

JeLee-river committed:
test: 기존의 useCheckboxHandler가 두 개의 파일로 분리된 점을 고려해 리팩토링

JeLee-river committed:
fix: 삭제된 장바구니 아이템 데이터가 로컬 스토리지에 반영되지 않은 문제 수정

JeLee-river committed:
refactor: useCheckboxHandler 로직의 변경에 따른 코드 수정

JeLee-river committed:
test: getStoredData 함수의 모킹 에러 개선

JeLee-river commented:
안녕하세요 쵸파! 첫 요청 때 구현하지 못한 요구사항과 피드백을 모두 반영 완료하여 재요청 드립니다! 빠르게 피드백을 주셨는데, 늦게 요청드려서 정말 죄송합니다. 😭 

아래는 요구 사항과 관련하여 새롭게 추가된 부분과 주요 변경 사항입니다!

---

<br>

## 요구 사항과 관련하여 추가된 목록

### 1️⃣ 로컬 스토리지

**useSelect 커스텀 훅에 로컬 스토리지 관리 기능 추가**

- 사용자의 선택을 로컬 스토리지에 저장할 수 있는 기능 추가 [(f11fccb)](https://github.com/woowacourse/react-shopping-cart/pull/407/commits/f11fccb9e06d538273774420b91565ccd20a909f)

- 초기 렌더링에서 전체 선택을 활성화하는 옵션을 true로 설정했을 때, 사용자가 선택 박스를 토글한 기록이 없는 아이템은 선택된 상태로 표시되도록하는 코드 추가 [(7a68f15)](https://github.com/woowacourse/react-shopping-cart/pull/407/commits/7a68f15634ff13516f1f3f7341dd69c67f1f821e)
  - **로직 설명**
  - 최초 렌더링 시, 모든 아이템이 선택된 상태로 표시된다. (단, 최대 선택 개수를 제한한 경우 제외)
  - 사용자가 아이템을 토글했을 때, 새로고침 시, 체크박스에 마지막 상태가 반영된 상태로 보인다.
  - 사용자가 아이템을 토글한 상태에서, 새로운 아이템을 장바구니에 넣은 경우, 해당 아이템은 선택된 상태로 표시된다.
  - 사용자가 아이템을 삭제한 후, 동일한 아이템을 다시 담을 시, 이전 선택 기록은 초기화된다. (새로운 아이템으로 간주)

<br>

### 2️⃣ 쿠폰

**2개 이상의 쿠폰 적용 시, 최고 할인 금액으로 적용하는 기능 구현**

- 쿠폰들을 적용하는 순서를 case별로 나누고, 각 case의 할인 금액을 계산하여 가장 높은 금액을 반환하도록 구현 [(877eabb)](https://github.com/woowacourse/react-shopping-cart/pull/407/commits/877eabbb75c8a7c276028d0780316054ec05dac4)

<br>

###  3️⃣ 테스트

**쿠폰 로직 테스트 추가**

- 쿠폰과 관련된 도메인 로직 테스트를 추가함 [(ffd7f44)](https://github.com/woowacourse/react-shopping-cart/pull/407/commits/ffd7f441de2910f2aacd981ef032fe66b38928ab)

  - **테스트 목록**
    - 최소 주문 금액이 존재하는 쿠폰들의 적용 가능 여부를 경계값으로 테스트
    - 최소 주문 수량이 존재하는 쿠폰들의 적용 가능 여부를 경계값으로 테스트
    - 사용 가능 시간이 정해져 있는 쿠폰의 적용 가능 여부를 경계값으로 테스트
    - 1개 이상의 쿠폰을 적용했을 때, 산정되는 쿠폰 할인 금액 테스트
    - 최적의 쿠폰 조합을 계산하는 로직 테스트
  

<br>
<br>

## 주요 변경 사항

### 1️⃣ 쿠폰 로직 추상화

**useCoupon 커스텀 훅을 제거하고, 쿠폰과 관련된 로직을 도메인으로 재분류**

- 모든 쿠폰들이 공통적으로 적용가능 여부, 할인 금액을 반환해야 하는 점을 고려하여, service 계층을 추가해 이 두가지 기능을 추상화함 [(25b0bd7)](https://github.com/woowacourse/react-shopping-cart/pull/407/commits/25b0bd737c0b4781f49f03a6105dc3c3280472e8)

- 쿠폰에 따라 다른 구현체를 적용하기 위해 사용하던 switch-case문을 제거하고, service의 함수를 호출하도록 함

<br>

### 2️⃣ 주문 상품 컴포넌트 생성

**OrderItem 컴포넌트 생성**

- 기존의 CartItem 컴포넌트를 재사용하던 주문 확인 페이지의 상품 UI를 OrderItem으로 대체함[(6de7868)](https://github.com/woowacourse/react-shopping-cart/pull/407/commits/6de7868cac9a7cd2861f9691032907482bb43b1a)

- CartItem을 장바구니 페이지와 주문 확인 페이지에서 모두 사용했을 때, 불필요한 props를 옵셔널로 전달받게 됨. 이에 따라 타입 안정성이 떨어지는 문제가 있어, 주문 확인 페이지에서 사용할 컴포넌트를 생성함

<br>

### 3️⃣ useCheckboxHandler 로직 분리

**useAllCheckboxHandler를 생성하여 전체 선택 로직을 관리하도록 함**

- useCheckboxHandler에서 모든 로직을 관리했을 때, props로 받는 option에 따라 return type이 달라지는 문제가 존재함

- useAllSelect 훅을 관리하는 로직을 useAllCheckboxHandler에서 관리하도록 함. 이에 따라 불필요한 option(enableAllSelect)을 제거하고 전체 선택이 필요한 곳에서 직접 호출하여 사용할 수 있도록 함 [(2857bdb)](https://github.com/woowacourse/react-shopping-cart/pull/407/commits/2857bdbcfbbea9d87eb0177bd645e9b7dae3353e)

<br>
<br>

## 기타 변경 사항

 - 현재 시각이 특정 시간 범위 내에 존재하는지 확인하는 유틸 생성[(c781c51)](https://github.com/woowacourse/react-shopping-cart/pull/407/commits/c781c5166f3a8690c5fdd1982011092b02a929d1)

- 주문 확인 페이지로 이동하는 기능의 테스트 코드 수정 [(81f6cb1)](https://github.com/woowacourse/react-shopping-cart/pull/407/commits/81f6cb163f24a1f60cdc4b73ceaf651611bb4fbb)

<br>
<br>

## 코멘트 답변

> 또는 제나가 특히 걱정되는 페이지나 컴포넌트가 있을까요?

제가 가장 걱정했던 곳은 주문 확인 페이지(OrderSummary) 입니다! 특히 OrderSummary의 경우, 쿠폰 모달에 필요한 props까지 생성하여 전달해주고 있기 때문에 지나치게 비대하지 않나 하는 의문이 들었습니다. 

또한 할인 금액을 계산하는 함수가 OrderSummary와 쿠폰 모달에서 중복으로 사용되다 보니, 책임 분리를 어떻게 해야 할 지 고민이 있었습니다. 😂 OrderSummary의 경우, 모달에서 적용할 쿠폰을 확정해 주어야 할인 금액을 계산할 수 있고, 쿠폰 모달은 사용자가 선택한 쿠폰을 기준으로 할인 금액을 계산하기 때문에, 이 미묘한 차이 때문에 더 복잡하다고 느낀 것 같습니다.

따라서 현재 OrderSummary 페이지 컴포넌트가 다른 컴포넌트나 훅에 위임할 수 있는 책임을 과도하게 지고 있지는 않은지 궁금합니다..! 

또한 일반적인 페이지 컴포넌트에서는,

1. 어디까지의 로직(API 호출, 도메인 로직 실행 등)을 직접 담고,

2. 어디부터 하위 컴포넌트나 훅에 위임하는 것이 유지보수와 재사용성 관점에서 좋을지, 구체적인 기준이 있다면 여쭙고 싶습니다.

---

감사합니다!

bassyu commented on src/pages/CouponModal/CouponModal.tsx:
```diff
@@ -0,0 +1,110 @@
+  setOpenModal: React.Dispatch<React.SetStateAction<boolean>>;
+  redeemCoupons: (coupons: CouponType[]) => void;
+  getDiscount: (selectedCoupons: CouponType[]) => number;
+  filterNonExpiredCoupons: (coupons: CouponType[]) => CouponType[];
```
> 제가 가장 걱정했던 곳은 주문 확인 페이지(OrderSummary) 입니다! 특히 OrderSummary의 경우, 쿠폰 모달에 필요한 props까지 생성하여 전달해주고 있기 때문에 지나치게 비대하지 않나 하는 의문이 들었습니다.

어느정도 분리 잘 해주신 것 같아요!

filterNonExpiredCoupons는 직접 import해도 괜찮을까요?

bassyu commented on src/pages/CouponModal/CouponModal.tsx:
쿠폰 모달이 열릴 때, 이전에 선택해둔 체크박스가 유지되면 좋겠습니다~!

쿠폰 자동 선택이 어느 타이밍에 적용되는지도 궁금합니다!

bassyu commented on src/pages/OrderSummary/OrderSummary.tsx:
항상 배송비가 무료인 이슈가 있습니다..!

<img width="100" alt="스크린샷 2025-06-14 오후 2 45 45" src="https://github.com/user-attachments/assets/89aafa0c-9e40-4ee1-b6e6-1d30926b43d0" />

bassyu commented on src/domain/coupon/couponHandler.ts:
```diff
@@ -0,0 +1,77 @@
+  [K in keyof CouponMapType]: HandlerType<CouponMapType[K]>;
+};
+
+const couponHandlerMap: CouponHandlerMapType = {
```
복잡한 쿠폰 로직에서 그룹화를 잘 해주셨네요! 👍 👍

bassyu reviewed (approved):
안녕하세요 제나!

관심사를 깔끔하게 잘 분리해주셨네요~! 가독성이 좋아서 코드를 읽기 용이했습니다! 👍 👍 
다만 요구사항과 앱의 퀄리티를 더 높은 우선순위로 작업해주시면 더욱 좋을 것 같습니다~!

2단계도 이만 머지하겠습니다~!
코멘트 남긴 부분은 이후에 개인적으로 보완 해주시면 좋을 것 같아요!
이번 미션도 수고 많으셨고, 다음 단계도 파이팅입니다~! 💪 💪
